src/engine/sequencer.js

```diff
@@ -155,12 +153,11 @@ class Sequencer {
                 let nextActiveThread = 0;
                 for (let i = 0; i < this.runtime.threads.length; i++) {
                     const thread = this.runtime.threads[i];
-                    if (thread.stack.length !== 0 &&
-                        thread.status !== Thread.STATUS_DONE) {
+                    if (thread.status === Thread.STATUS_DONE) {
+                        doneThreads.push(thread);
+                    } else {
                         this.runtime.threads[nextActiveThread] = thread;
                         nextActiveThread++;
-                    } else {
-                        doneThreads.push(thread);
                     }
                 }
                 this.runtime.threads.length = nextActiveThread;
```

Comment:
Now here's the interesting part. This is an "old fashioned" filter which overwrites the array in-place as it iterates over it (it'll always be done with the current item before that item gets overwritten), setting `.length` at the end to "chop off" the garbage items. It's semantically the same as:

```js
doneThreads.push(
  ...this.runtime.threads.filter(({status}) => status === Thread.STATUS_DONE));

this.runtime.threads =
     this.runtime.threads.filter(({status}) => status !== Thread.STATUS_DONE);
```

(Try running through the loop in your head if you need to convince yourself of this!)

So let's rewind to `activeThreads`. Remember that it's set to true if any thread's status *isn't* `STATUS_DONE`.

Now, here's the `while` loop at the very top level of `stepThreads`:

```js
while (this.runtime.threads.length > 0 &&
       activeThreads &&
       /* extra conditions */) {...}
```

What I've been wanting to figure out in the review of this commit is, is it necessary to check `this.runtime.threads.length > 0` here *at all?* Or does `activeThreads` already contain that information for us? Assuming that this loop always iterates at least once (for the moment).

Well the `this.runtime.threads.length = nextActiveThread` here kind of gives it away. It and `activeThreads` sure *sound* synonymous. Is that so? Yes — `threads.length` will be non-zero iff any thread is not `STATUS_DONE`; just the same for `activeThreads`.

Note that this filter runs *after* we've iterated over all active threads - it's the last step in the top-level `stepThreads` loop. If it were somehow possible for a thread's status to be changed to `STATUS_DONE`, *after* it was the one responsible for `activeThreads` becoming true, then there would be trouble — this filter could resolve to zero (no active threads) while `activeThreads` remains true (because nothing sets it to false!). But there just isn't any opportunity for a thread's statuses to become `STATUS_DONE` — the only thing setting its status to `STATUS_DONE` is inside `stepThread(activeThread)`.

So yes, `this.runtime.threads.length > 0` really does become synonymous with `activeThreads` after the first iteration.

<hr>

Now for the takeaway! To simplify this logic, how about we only check `this.runtime.threads.length > 0` on the first iteration? A simple way to do this would be to just change the initial definition for `activeThreads`:

```diff
-let activeThreads = true;
+let activeThreads = this.runtime.threads.length > 0;
```

Et voila: the `while` loop will still neatly not perform its first iteration, because `activeThreads` was appropriately false, and we skip ahead.

But what exactly are we skipping *to?* Well, we set `this.activeThread = null` and return `doneThreads` below - that's everything. If it's possible, it would be cleaner to simply use an early return. Here's the naive example:

```js
if (this.runtime.threads.length === 0) {
    this.activeThread = null;
    return doneThreads;
}
```

That would work, but we'd like to determine the results of those two steps if the `while` loop was outright skipped - if we do, we might be able to simplify this early return, or move it higher up in the function (instead of right before the main `while` loop).

So for the first line: `this.activeThread = null` is purely a clean-up line. In other words it guarantees that `activeThread` is null once `stepThreads` exits, and so only has a value for the duration of `stepThreads`.

But `stepThreads` is also the only code *setting* `activeThread`... so we know that if it *was* null *before* any code evaluated, it doesn't need to be reset to null during the early return either.

```js
if (this.runtime.threads.length === 0) {
    return doneThreads;
}
```

*(Note: `activeThread` was outright made into a local variable later, with the mentioned cleanup removed. Of course, cleanup remains unneeded in the early return.)*

What's that `doneThreads` value going to be? Not much surprise here: it's just going to be an empty array! The only code which pushes to it is inside the main `while` loop, so if the `while` loop is skipped, nothing gets put in it.

```js
if (this.runtime.threads.length === 0) {
    return [];
}
```

This can be moved *all the way to the top* of `stepThreads`. (Or past the side effects `this.timer.start()` and `this.runtime.updateCurrentMSecs()`, if those have any impact outside of `stepThreads`. And only if!)

An early return for the "no active threads" case means we *know* we're going to run the thread loop at least once. So the final clean-up I'd do (while keeping the general structure of `stepThreads` the same) is to flat-out replace the `while` loop with a `do`-`while`:

```js
let activeThreads;
do {
    activeThreads = false;
    /* all the usual logic */
} while (activeThreads &&
         this.timer.timeElapsed() < WORK_TIME &&
         (this.runtime.turboMode || !this.runtime.redrawRequested))
```

Yes, it's true — `do`-`while` loops *are* spooky scary skeletons. But it really does make sense to move this conditional to the end, even beyond the semantic fact that it's "fine" to skip it the very first time — see, the only case where the conditional could *possibly* be true is *after* the some an iteration is complete! That's clear as much when you actually read the conditional parts:

* There only might *not* be `activeThreads` after we've evaluated all the were-active threads and they turned out to become done.
* The work timer could only possibly have elapsed if, you know, we've actually done any work.
* A redraw could only have been requested if some block requested it... as a result of... stepping a thread!

I haven't read ahead so who knows how much `stepThreads` is going to be restructured further. But those are my suggestions based on its current state, and thinking of how I'd refactor if I were the one doing the PR!